# Toronto Postal Codes

### Import Libraries

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

import geocoder

#!conda install -c conda-forge geopy --yes
from geopy.geocoders import Nominatim

#!conda install -c conda-forge folium --yes
import folium

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

print('Libraries loaded.')

Libraries loaded.


## Question #1

### 1.1 Retrieve data

In [2]:
# get data from the Wikipedia page
list_of_dfs = pd.read_html('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M')

toronto_df = list_of_dfs[0]
#toronto_df = toronto_df.rename(columns={'Postal Code' : 'PostalCode'})
toronto_df.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

### 1.2 Remove Not Assigned Boroughs

In [3]:
toronto_df.drop(toronto_df[toronto_df['Borough'] == 'Not assigned'].index, inplace = True) 
toronto_df.reset_index(inplace=True, drop=True)
toronto_df.head()

Postal Code           Borough                                Neighbourhood
0         M3A        North York                                    Parkwoods
1         M4A        North York                             Victoria Village
2         M5A  Downtown Toronto                    Regent Park, Harbourfront
3         M6A        North York             Lawrence Manor, Lawrence Heights
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

### 1.3 Check Not Assigned Neighbourhood

In [4]:
toronto_df[toronto_df['Neighbourhood'] == 'Not assigned']

Empty DataFrame
Columns: [Postal Code, Borough, Neighbourhood]
Index: []

There is NO **Not Assigned** Neighbourhood

### 1.4 Final result. Dataframe Shape

In [5]:
toronto_df.shape

(103, 3)

The dataframe hase **103 rows** and **3 columns**

## Question #2

### Populate the Latitude and Longitude of postal code dataframe

In [6]:
def get_location_remote(row) :
    # initialize your variable to None
    lat_lng_coords = None
    while(lat_lng_coords is None):
        g = geocoder.google('{}, Toronto, Ontario'.format(row['Postal Code']))
        lat_lng_coords = g.latlng

    row['Latitude'] = lat_lng_coords[0]
    row['Longitude'] = lat_lng_coords[1]
    
    return row

# Use CSV
toronto_geo = pd.read_csv('https://cocl.us/Geospatial_data')
toronto_df = toronto_df.merge(toronto_geo, on='Postal Code')
toronto_df.head()

# User geocoder
#toronto_df['Latitude', 'Longitude'] = 
#toronto_df = toronto_df.apply(get_location_remote, axis=1)
#toronto_df.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

## Question #3

Quickly examine the resulting dataframe.

In [7]:
toronto_df.head()

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   

    Latitude  Longitude  
0  43.753259 -79.329656  
1  43.725882 -79.315572  
2  43.654260 -79.360636  
3  43.718518 -79.464763  
4  43.662301 -79.389494

In [8]:
print('The dataframe has {} boroughs and {} neighborhoods.'.format(
        len(toronto_df['Borough'].unique()),
        toronto_df.shape[0]
    )
)

The dataframe has 10 boroughs and 103 neighborhoods.


#### Use geopy library to get the latitude and longitude values of Toronto City.


In [9]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto City are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto City are 43.6534817, -79.3839347.


#### Create a map of Toronto with neighborhoods superimposed on top.


In [10]:
# create map of New York using latitude and longitude values
map_newyork = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_df['Latitude'], toronto_df['Longitude'], toronto_df['Borough'], toronto_df['Neighbourhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_newyork)  
    
map_newyork

Next, we are going to start utilizing the Foursquare API to explore the neighborhoods and segment them.


#### Define Foursquare Credentials and Version


In [11]:
CLIENT_ID = 'Y0P44AF55JQIYQ5Q24KAQ0X4MQRM0LCUAPSZ2OQ4MKL2A1GA' # your Foursquare ID
CLIENT_SECRET = 'BW5TXHHN1QJNA5T3EHJTOPV2GDDBJAVG1WAGAL2MWR0WB01S' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version
LIMIT = 100 # A default Foursquare API limit value

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: Y0P44AF55JQIYQ5Q24KAQ0X4MQRM0LCUAPSZ2OQ4MKL2A1GA
CLIENT_SECRET:BW5TXHHN1QJNA5T3EHJTOPV2GDDBJAVG1WAGAL2MWR0WB01S


### Explore Neighborhoods in Toronto

#### Let's create a function to repeat the same process to all the neighborhoods in Toronto


In [12]:
def getNearbyVenues(post_codes, boroughs, neighbourhoods, latitudes, longitudes, radius=500):
    venues_list=[]
    for post_code, borough, neighborhood, lat, lng in zip(post_codes, boroughs, neighbourhoods, latitudes, longitudes):
        print(post_code)

        # create the API request URL and params
        url = 'https://api.foursquare.com/v2/venues/explore'    
        params = dict(
        client_id=CLIENT_ID,
        client_secret=CLIENT_SECRET,
        v=VERSION,
        ll='{}, {}'.format(lat, lng),
        limit=LIMIT,
        radius=radius
        )
        
            
        # make the GET request
        results = requests.get(url=url, params=params).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            post_code, 
            borough,
            neighborhood,
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Postal Code', 
                  'Borough',
                  'Neighbourhood',
                  'Postal Code Latitude', 
                  'Postal Code Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [13]:
# type your answer here
toronto_venues = getNearbyVenues(toronto_df['Postal Code'], toronto_df['Borough'], toronto_df['Neighbourhood'], toronto_df['Latitude'], toronto_df['Longitude'], radius=2000)


M3A
M4A
M5A
M6A
M7A
M9A
M1B
M3B
M4B
M5B
M6B
M9B
M1C
M3C
M4C
M5C
M6C
M9C
M1E
M4E
M5E
M6E
M1G
M4G
M5G
M6G
M1H
M2H
M3H
M4H
M5H
M6H
M1J
M2J
M3J
M4J
M5J
M6J
M1K
M2K
M3K
M4K
M5K
M6K
M1L
M2L
M3L
M4L
M5L
M6L
M9L
M1M
M2M
M3M
M4M
M5M
M6M
M9M
M1N
M2N
M3N
M4N
M5N
M6N
M9N
M1P
M2P
M4P
M5P
M6P
M9P
M1R
M2R
M4R
M5R
M6R
M7R
M9R
M1S
M4S
M5S
M6S
M1T
M4T
M5T
M1V
M4V
M5V
M8V
M9V
M1W
M4W
M5W
M8W
M9W
M1X
M4X
M5X
M8X
M4Y
M7Y
M8Y
M8Z


#### Let's check the size of the resulting dataframe


In [14]:
print(toronto_venues.shape)
toronto_venues.head()

(8599, 9)


Postal Code     Borough Neighbourhood  Postal Code Latitude  \
0         M3A  North York     Parkwoods             43.753259   
1         M3A  North York     Parkwoods             43.753259   
2         M3A  North York     Parkwoods             43.753259   
3         M3A  North York     Parkwoods             43.753259   
4         M3A  North York     Parkwoods             43.753259   

   Postal Code Longitude                        Venue  Venue Latitude  \
0             -79.329656              Allwyn's Bakery       43.759840   
1             -79.329656  Donalda Golf & Country Club       43.752816   
2             -79.329656                 Island Foods       43.745866   
3             -79.329656           Graydon Hall Manor       43.763923   
4             -79.329656                  Tim Hortons       43.760668   

   Venue Longitude        Venue Category  
0       -79.324719  Caribbean Restaurant  
1       -79.342741           Golf Course  
2       -79.346035  Caribbean Restaurant  
3       -79.342961           Event Space  
4       -79.326368                  Café

Let's check how many venues were returned for each neighborhood


In [15]:
toronto_venues.groupby('Postal Code').count()

Borough  Neighbourhood  Postal Code Latitude  \
Postal Code                                                 
M1B               45             45                    45   
M1C               39             39                    39   
M1E               39             39                    39   
M1G               64             64                    64   
M1H              100            100                   100   
M1J               76             76                    76   
M1K               58             58                    58   
M1L               90             90                    90   
M1M               40             40                    40   
M1N               44             44                    44   
M1P               93             93                    93   
M1R               98             98                    98   
M1S              100            100                   100   
M1T               82             82                    82   
M1V              100            100                   100   
M1W               76             76                    76   
M1X                5              5                     5   
M2H               75             75                    75   
M2J              100            100                   100   
M2K               54             54                    54   
M2L              100            100                   100   
M2M              100            100                   100   
M2N               41             41                    41   
M2P              100            100                   100   
M2R               52             52                    52   
M3A               91             91                    91   
M3B               96             96                    96   
M3C               93             93                    93   
M3H               51             51                    51   
M3J               88             88                    88   
M3K              100            100                   100   
M3L               25             25                    25   
M3M               57             57                    57   
M3N               48             48                    48   
M4A              100            100                   100   
M4B               52             52                    52   
M4C               99             99                    99   
M4E              100            100                   100   
M4G               86             86                    86   
M4H              100            100                   100   
M4J              100            100                   100   
M4K              100            100                   100   
M4L              100            100                   100   
M4M              100            100                   100   
M4N               93             93                    93   
M4P              100            100                   100   
M4R              100            100                   100   
M4S              100            100                   100   
M4T              100            100                   100   
M4V              100            100                   100   
M4W              100            100                   100   
M4X              100            100                   100   
M4Y              100            100                   100   
M5A              100            100                   100   
M5B              100            100                   100   
M5C              100            100                   100   
M5E              100            100                   100   
M5G              100            100                   100   
M5H              100            100                   100   
M5J              100            100                   100   
M5K              100            100                   100   
M5L              100            100                   100   
M5M               98             98                    98   
M5N              100            100             

#### Let's find out how many unique categories can be curated from all the returned venues


In [16]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 329 uniques categories.


### Analyze Each Neighborhood


In [17]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Postal Code'] = toronto_venues['Postal Code'] 

# move neighborhood column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

Gas Station  Gastropub  Gay Bar  General Entertainment  German Restaurant  \
0            0          0        0                      0                  0   
1            0          0        0                      0                  0   
2            0          0        0                      0                  0   
3            0          0        0                      0                  0   
4            0          0        0                      0                  0   

   Gift Shop  Golf Course  Gourmet Shop  Greek Restaurant  Grocery Store  Gym  \
0          0            0             0                 0              0    0   
1          0            1             0                 0              0    0   
2          0            0             0                 0              0    0   
3          0            0             0                 0              0    0   
4          0            0             0                 0              0    0   

   Gym / Fitness Center  Gym Pool  Gymnastics Gym  Hakka Restaurant  \
0                     0         0               0                 0   
1                     0         0               0                 0   
2                     0         0               0                 0   
3                     0         0               0                 0   
4                     0         0               0                 0   

   Harbor / Marina  Hardware Store  Hawaiian Restaurant  \
0                0               0                    0   
1                0               0                    0   
2                0               0                    0   
3                0               0                    0   
4                0               0                    0   

   Health & Beauty Service  Health Food Store  High School  Historic Site  \
0                        0                  0            0              0   
1                        0                  0            0              0   
2                        0                  0            0              0   
3                        0                  0            0              0   
4                        0                  0            0              0   

   History Museum  Hobby Shop  Hockey Arena  Home Service  \
0               0           0             0             0   
1               0           0             0             0   
2               0           0             0             0   
3               0           0             0             0   
4               0           0             0             0   

   Hong Kong Restaurant  Hookah Bar  Hostel  Hotel  Hotel Bar  \
0                     0           0       0      0          0   
1                     0           0       0      0          0   
2                     0           0       0      0          0   
3                     0           0       0      0          0   
4                     0           0       0      0          0   

   Hotpot Restaurant  Hungarian Restaurant  Ice Cream Shop  Indian Restaurant  \
0                  0                     0               0                  0   
1                  0                     0               0                  0   
2                  0                     0               0                  0   
3                  0                     0               0                  0   
4                  0                     0               0                  0   

   Indie Movie Theater  Indonesian Restaurant  Indoor Play Area  Intersection  \
0                    0                      0                 0             0   
1                    0                      0                 0             0   
2                    0                      0                 0             0   
3                    0                      0                 0             0   
4                    0                      0                 0             0   

   Italian Restaurant  Japanese Restaurant  Jewelry Store 

And let's examine the new dataframe size.

In [18]:
toronto_onehot.shape

(8599, 330)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category


In [19]:
toronto_grouped = toronto_onehot.groupby('Postal Code').mean().reset_index()
toronto_grouped

000000          0.000000   
93              0.00                       0.000000          0.000000   
94              0.00                       0.000000          0.000000   
95              0.00                       0.000000          0.000000   
96              0.00                       0.000000          0.000000   
97              0.00                       0.000000          0.000000   
98              0.00                       0.000000          0.000000   
99              0.00                       0.000000          0.000000   
100             0.00                       0.000000          0.000000   
101             0.00                       0.000000          0.000000   
102             0.00                       0.000000          0.000000   

     Vietnamese Restaurant  Volleyball Court  Warehouse Store  Whisky Bar  \
0                 0.000000              0.00         0.000000        0.00   
1                 0.000000              0.00         0.000000        0.00   
2                 0.000000              0.00         0.000000        0.00   
3                 0.015625              0.00         0.000000        0.00   
4                 0.010000              0.00         0.000000        0.00   
5                 0.000000              0.00         0.000000        0.00   
6                 0.000000              0.00         0.017241        0.00   
7                 0.011111              0.00         0.000000        0.00   
8                 0.000000              0.00         0.000000        0.00   
9                 0.000000              0.00         0.000000        0.00   
10                0.010753              0.00         0.010753        0.00   
11                0.020408              0.00         0.010204        0.00   
12                0.020000              0.00         0.000000        0.00   
13                0.012195              0.00         0.000000        0.00   
14                0.040000              0.00         0.000000        0.00   
15                0.026316              0.00         0.000000        0.00   
16                0.000000              0.00         0.000000        0.00   
17                0.000000              0.00         0.013333        0.00   
18                0.000000              0.00         0.000000        0.00   
19                0.000000              0.00         0.000000        0.00   
20                0.000000              0.00         0.000000        0.00   
21                0.000000              0.00         0.000000        0.00   
22                0.000000              0.00         0.000000        0.00   
23                0.000000              0.00         0.000000        0.00   
24                0.019231              0.00         0.000000        0.00   
25                0.000000              0.00         0.000000        0.00   
26                0.000000              0.00         0.000000        0.00   
27                0.000000              0.00         0.000000        0.00   
28                0.000000              0.00         0.000000        0.00   
29                0.022727              0.00         0.000000        0.00   
30                0.030000              0.00         0.010000        0.00   
31                0.120000              0.00         0.000000        0.00   
32                0.087719              0.00         0.000000        0.00   
33                0.041667              0.00         0.000000        0.00   
34                0.000000              0.01         0.000000        0.00   
35                0.000000              0.00         0.000000        0.00   
36                0.000000              0.00         0.000000        0.00   
37                0.000000              0.00         0.000000        0.00   
38                0.000000              0.00         0.011628        0.00   
39                0.000000              0.00         0.010000        0.00   
40                0.000000              0.00         0.000000        0.00   
41     

#### Let's confirm the new size

In [20]:
toronto_grouped.shape

(103, 330)

#### Let's print each neighborhood along with the top 5 most common venues


In [21]:
num_top_venues = 5

for hood in toronto_grouped['Postal Code']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Postal Code'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----M1B----
                  venue  freq
0           Zoo Exhibit  0.33
1  Fast Food Restaurant  0.07
2           Pizza Place  0.04
3                   Zoo  0.04
4           Gas Station  0.04


----M1C----
            venue  freq
0     Coffee Shop  0.08
1  Breakfast Spot  0.08
2            Park  0.05
3            Bank  0.05
4  Sandwich Place  0.05


----M1E----
                  venue  freq
0           Pizza Place  0.13
1                  Bank  0.05
2         Train Station  0.05
3  Fast Food Restaurant  0.05
4           Coffee Shop  0.05


----M1G----
                  venue  freq
0           Coffee Shop  0.14
1  Fast Food Restaurant  0.09
2        Sandwich Place  0.06
3        Discount Store  0.05
4           Pizza Place  0.05


----M1H----
            venue  freq
0     Coffee Shop  0.10
1     Gas Station  0.06
2  Clothing Store  0.05
3      Restaurant  0.05
4  Sandwich Place  0.04


----M1J----
                  venue  freq
0           Coffee Shop  0.09
1  Fast Food Restaurant  0.09


#### Let's put that into a _pandas_ dataframe
First, let's write a function to sort the venues in descending order.



In [22]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.


In [23]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Postal Code']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
toronto_venues_sorted = pd.DataFrame(columns=columns)
toronto_venues_sorted['Postal Code'] = toronto_grouped['Postal Code']

for ind in np.arange(toronto_grouped.shape[0]):
    toronto_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

toronto_venues_sorted.head()

Postal Code 1st Most Common Venue 2nd Most Common Venue  \
0         M1B           Zoo Exhibit  Fast Food Restaurant   
1         M1C           Coffee Shop        Breakfast Spot   
2         M1E           Pizza Place        Breakfast Spot   
3         M1G           Coffee Shop  Fast Food Restaurant   
4         M1H           Coffee Shop           Gas Station   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0                   Zoo    Athletics & Sports           Gas Station   
1           Pizza Place              Pharmacy             Pet Store   
2                  Bank  Fast Food Restaurant         Train Station   
3        Sandwich Place        Discount Store           Pizza Place   
4        Clothing Store            Restaurant        Sandwich Place   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0            Restaurant           Pizza Place   Fried Chicken Joint   
1                  Park        Sandwich Place                  Bank   
2           Coffee Shop           Supermarket                   Bar   
3                  Bank           Supermarket            Beer Store   
4  Fast Food Restaurant                   Gym      Department Store   

  9th Most Common Venue 10th Most Common Venue  
0          Skating Rink   Caribbean Restaurant  
1        Ice Cream Shop           Burger Joint  
2      Greek Restaurant             Beer Store  
3     Indian Restaurant     Chinese Restaurant  
4              Pharmacy                   Bank

### Cluster Neighborhoods


Run _k_-means to cluster the neighborhood into 5 clusters.


In [24]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Postal Code', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 2, 2, 2, 1, 2, 2, 2, 2, 2], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.


In [25]:
# add clustering labels
toronto_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_df

# merge toronto_merged with toronto_df to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(toronto_venues_sorted.set_index('Postal Code'), on='Postal Code')

toronto_merged.head(10) # check the last columns!

Postal Code           Borough                                Neighbourhood  \
0         M3A        North York                                    Parkwoods   
1         M4A        North York                             Victoria Village   
2         M5A  Downtown Toronto                    Regent Park, Harbourfront   
3         M6A        North York             Lawrence Manor, Lawrence Heights   
4         M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
5         M9A         Etobicoke      Islington Avenue, Humber Valley Village   
6         M1B       Scarborough                               Malvern, Rouge   
7         M3B        North York                                    Don Mills   
8         M4B         East York              Parkview Hill, Woodbine Gardens   
9         M5B  Downtown Toronto                     Garden District, Ryerson   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.753259 -79.329656               1           Coffee Shop   
1  43.725882 -79.315572               2           Coffee Shop   
2  43.654260 -79.360636               0           Coffee Shop   
3  43.718518 -79.464763               1        Clothing Store   
4  43.662301 -79.389494               0           Coffee Shop   
5  43.667856 -79.532242               1              Pharmacy   
6  43.806686 -79.194353               4           Zoo Exhibit   
7  43.745906 -79.352188               1           Coffee Shop   
8  43.706397 -79.309937               2           Pizza Place   
9  43.657162 -79.378937               0           Coffee Shop   

  2nd Most Common Venue   3rd Most Common Venue 4th Most Common Venue  \
0    Chinese Restaurant     Japanese Restaurant           Gas Station   
1  Fast Food Restaurant                     Gym        Sandwich Place   
2                  Park              Restaurant                  Café   
3           Coffee Shop  Furniture / Home Store  Fast Food Restaurant   
4                 Hotel                    Café   Japanese Restaurant   
5           Coffee Shop                    Bank         Grocery Store   
6  Fast Food Restaurant                     Zoo    Athletics & Sports   
7            Restaurant                    Bank                  Park   
8                  Park            Skating Rink        Sandwich Place   
9                  Café              Restaurant             Gastropub   

  5th Most Common Venue      6th Most Common Venue 7th Most Common Venue  \
0        Discount Store                   Pharmacy   Fried Chicken Joint   
1   Japanese Restaurant  Middle Eastern Restaurant        Clothing Store   
2                 Diner                        Pub        Farmers Market   
3         Grocery Store                 Restaurant                  Bank   
4      Ramen Restaurant                Pizza Place          Dance Studio   
5                  Park                       Café           Golf Course   
6           Gas Station                 Restaurant           Pizza Place   
7           Pizza Place        Japanese Restaurant           Supermarket   
8           Coffee Shop             Ice Cream Shop         Grocery Store   
9   Japanese Restaurant                Pizza Place      Ramen Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0         Sandwich Place           Supermarket            Pizza Place  
1          Grocery Store           Gas Station       Department Store  
2              Gastropub                Bakery    Japanese Restaurant  
3  Vietnamese Restaurant          Dessert Shop       Greek Restaurant  
4                   Park      Sushi Restaurant            Supermarket  
5           Liquor Store           Pizza Place          Shopping Mall  
6    Fried Chicken Joint          Skating Rink   Caribbean Restaurant  
7           Burger Joint    Italian Restaurant                   Café  
8   Fast Food Restaurant            Beer Store                Brewery  
9        Thai Restaurant                  Pa

Finally, let's visualize the resulting clusters


In [26]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighbourhood'], toronto_merged['Cluster Labels'].astype(int)):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Examine Clusters

Now, you can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, you can then assign a name to each cluster. I will leave this exercise to you.


#### Cluster 1


In [27]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
2    Downtown Toronto               0           Coffee Shop   
4    Downtown Toronto               0           Coffee Shop   
9    Downtown Toronto               0           Coffee Shop   
14          East York               0                  Park   
15   Downtown Toronto               0           Coffee Shop   
16               York               0           Coffee Shop   
19       East Toronto               0           Coffee Shop   
20   Downtown Toronto               0           Coffee Shop   
23          East York               0           Coffee Shop   
24   Downtown Toronto               0           Coffee Shop   
25   Downtown Toronto               0                  Café   
30   Downtown Toronto               0           Coffee Shop   
31       West Toronto               0                  Café   
35          East York               0                  Café   
36   Downtown Toronto               0                 Hotel   
37       West Toronto               0                  Café   
41       East Toronto               0                  Café   
42   Downtown Toronto               0                  Café   
43       West Toronto               0           Coffee Shop   
47       East Toronto               0           Coffee Shop   
48   Downtown Toronto               0                  Café   
54       East Toronto               0           Coffee Shop   
55         North York               0           Coffee Shop   
59         North York               0     Korean Restaurant   
61    Central Toronto               0           Coffee Shop   
62    Central Toronto               0           Coffee Shop   
63               York               0           Coffee Shop   
67    Central Toronto               0           Coffee Shop   
68    Central Toronto               0    Italian Restaurant   
69       West Toronto               0           Coffee Shop   
73    Central Toronto               0    Italian Restaurant   
74    Central Toronto               0           Coffee Shop   
75       West Toronto               0           Coffee Shop   
79    Central Toronto               0    Italian Restaurant   
80   Downtown Toronto               0                  Café   
81       West Toronto               0                Bakery   
83    Central Toronto               0    Italian Restaurant   
84   Downtown Toronto               0                  Café   
86    Central Toronto               0    Italian Restaurant   
87   Downtown Toronto               0                  Park   
88          Etobicoke               0                  Park   
91   Downtown Toronto               0           Coffee Shop   
92   Downtown Toronto               0           Coffee Shop   
96   Downtown Toronto               0           Coffee Shop   
97   Downtown Toronto               0            Restaurant   
98          Etobicoke               0           Coffee Shop   
99   Downtown Toronto               0           Coffee Shop   
100      East Toronto               0           Coffee Shop   
101         Etobicoke               0           Coffee Shop   

    2nd Most Common Venue 3rd Most Common Venue      4th Most Common Venue  \
2                    Park            Restaurant                       Café   
4                   Hotel                  Café        Japanese Restaurant   
9                    Café            Restaurant                  Gastropub   
14            Coffee Shop           Pizza Place                       Café   
15                   Café            Restaurant                      Hotel   
16     Italian Restaurant                  Bank                       Café   
19                    Pub                 Beach             Breakfast Spot   
20                   Café                 Hotel                 Restaurant   
23      Indian Restaurant                Bakery                       Park   
24                   Café           Pizza Place                 Restaurant   
25   

#### Cluster 2


In [28]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0    North York               1           Coffee Shop    Chinese Restaurant   
3    North York               1        Clothing Store           Coffee Shop   
5     Etobicoke               1              Pharmacy           Coffee Shop   
7    North York               1           Coffee Shop            Restaurant   
10   North York               1        Clothing Store           Coffee Shop   
13   North York               1           Coffee Shop            Restaurant   
17    Etobicoke               1           Coffee Shop           Golf Course   
26  Scarborough               1           Coffee Shop           Gas Station   
27   North York               1           Coffee Shop                  Bank   
28   North York               1           Coffee Shop                  Park   
33   North York               1           Coffee Shop              Pharmacy   
34   North York               1           Coffee Shop        Sandwich Place   
39   North York               1                  Park    Chinese Restaurant   
40   North York               1        Clothing Store           Coffee Shop   
45   North York               1           Coffee Shop                  Bank   
52   North York               1     Korean Restaurant           Coffee Shop   
66   North York               1           Coffee Shop                  Bank   
76  Mississauga               1           Coffee Shop        Sandwich Place   
77    Etobicoke               1           Coffee Shop        Sandwich Place   
78  Scarborough               1    Chinese Restaurant           Coffee Shop   
85  Scarborough               1    Chinese Restaurant           Coffee Shop   
94    Etobicoke               1           Coffee Shop                 Hotel   

        3rd Most Common Venue      4th Most Common Venue  \
0         Japanese Restaurant                Gas Station   
3      Furniture / Home Store       Fast Food Restaurant   
5                        Bank              Grocery Store   
7                        Bank                       Park   
10     Furniture / Home Store                       Bank   
13        Japanese Restaurant                       Park   
17                       Bank              Grocery Store   
26             Clothing Store                 Restaurant   
27         Chinese Restaurant             Sandwich Place   
28                Pizza Place                   Pharmacy   
33  Middle Eastern Restaurant             Clothing Store   
34     Furniture / Home Store         Chinese Restaurant   
39                Coffee Shop                       Bank   
40         Athletics & Sports                 Food Court   
45                Pizza Place                       Café   
52            Bubble Tea Shop  Middle Eastern Restaurant   
66             Sandwich Place              Grocery Store   
76                      Hotel  Middle Eastern Restaurant   
77                 Restaurant                   Pharmacy   
78                 Restaurant                   Pharmacy   
85            Bubble Tea Shop                     Bakery   
94                Gas Station        American Restaurant   

   5th Most Common Venue      6th Most Common Venue  7th Most Common Venue  \
0         Discount Store                   Pharmacy    Fried Chicken Joint   
3          Grocery Store                 Restaurant                   Bank   
5                   Park                       Café            Golf Course   
7            Pizza Place        Japanese Restaurant            Supermarket   
10   Fried Chicken Joint       Fast Food Restaurant             Restaurant   
13           Pizza Place  Middle Eastern Restaurant                    Gym   
17        Baseball Field         College Rec Center            Gas Station   
26        Sandwich Place       Fast Food Restaurant                    Gym   
27              Pharmacy           Sushi Restaurant                   Park   
28         Deli / Bodega     

#### Cluster 3


In [29]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels  1st Most Common Venue  \
1     North York               2            Coffee Shop   
8      East York               2            Pizza Place   
11     Etobicoke               2            Coffee Shop   
12   Scarborough               2            Coffee Shop   
18   Scarborough               2            Pizza Place   
21          York               2     Italian Restaurant   
22   Scarborough               2            Coffee Shop   
29     East York               2                   Park   
32   Scarborough               2   Fast Food Restaurant   
38   Scarborough               2     Chinese Restaurant   
44   Scarborough               2   Fast Food Restaurant   
46    North York               2  Vietnamese Restaurant   
49    North York               2   Fast Food Restaurant   
50    North York               2            Coffee Shop   
51   Scarborough               2                   Park   
53    North York               2            Coffee Shop   
56          York               2            Coffee Shop   
57    North York               2          Grocery Store   
58   Scarborough               2            Coffee Shop   
60    North York               2            Coffee Shop   
64          York               2            Coffee Shop   
65   Scarborough               2            Coffee Shop   
70     Etobicoke               2            Pizza Place   
71   Scarborough               2            Coffee Shop   
72    North York               2            Coffee Shop   
82   Scarborough               2   Fast Food Restaurant   
89     Etobicoke               2            Coffee Shop   
90   Scarborough               2     Chinese Restaurant   
93     Etobicoke               2            Coffee Shop   
102    Etobicoke               2            Coffee Shop   

     2nd Most Common Venue      3rd Most Common Venue  \
1     Fast Food Restaurant                        Gym   
8                     Park               Skating Rink   
11       Convenience Store                       Park   
12          Breakfast Spot                Pizza Place   
18          Breakfast Spot                       Bank   
21          Sandwich Place                       Bank   
22    Fast Food Restaurant             Sandwich Place   
29             Coffee Shop             Sandwich Place   
32             Coffee Shop                Pizza Place   
38    Fast Food Restaurant              Grocery Store   
44             Coffee Shop             Sandwich Place   
46                    Park                Pizza Place   
49             Coffee Shop      Vietnamese Restaurant   
50                    Bank           Asian Restaurant   
51         Harbor / Marina                Coffee Shop   
53   Vietnamese Restaurant                       Bank   
56    Fast Food Restaurant     Furniture / Home Store   
57             Coffee Shop                 Beer Store   
58           Grocery Store                        Gym   
60             Gas Station                Pizza Place   
64          Sandwich Place     Furniture / Home Store   
65    Fast Food Restaurant                Pizza Place   
70           Grocery Store             Sandwich Place   
71              Restaurant  Middle Eastern Restaurant   
72             Pizza Place                       Park   
82                    Bank                Coffee Shop   
89             Pizza Place          Indian Restaurant   
90             Coffee Shop       Fast Food Restaurant   
93        Department Store       Fast Food Restaurant   
102             Restaurant       Fast Food Restaurant   

         4th Most Common Venue 5th Most Common Venue  \
1               Sandwich Place   Japanese Restaurant   
8               Sandwich Place           Coffee Shop   
11               Grocery Store           Pizza Place   
12                    Pharmacy             Pet Store   
18        Fast Food Restaurant         Train Station   
21      Furniture / Home Store                Bakery   
22              Dis

#### Cluster 4


In [30]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
95  Scarborough               3                  Farm           Golf Course   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
95      Sculpture Garden                 Trail            Playground   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
95           Zoo Exhibit     Electronics Store           Escape Room   

   9th Most Common Venue 10th Most Common Venue  
95  Ethiopian Restaurant          Event Service

#### Cluster 5

In [31]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
6  Scarborough               4           Zoo Exhibit  Fast Food Restaurant   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
6                   Zoo    Athletics & Sports           Gas Station   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
6            Restaurant           Pizza Place   Fried Chicken Joint   

  9th Most Common Venue 10th Most Common Venue  
6          Skating Rink   Caribbean Restaurant